consensus/fork_choice/src/fork_choice.rs

```diff
@@ -1425,15 +1445,56 @@ where
             .map_err(Into::into)
     }
 
+    /// Instantiate `Self` from some `PersistedForkChoice` generated by a earlier call to
+    /// `Self::to_persisted`.
+    pub fn proto_array_from_persisted(
+        persisted: &PersistedForkChoice,
+        reset_payload_statuses: ResetPayloadStatuses,
+        spec: &ChainSpec,
+        log: &Logger,
+    ) -> Result<ProtoArrayForkChoice, Error<T::Error>> {
+        let mut proto_array = ProtoArrayForkChoice::from_bytes(&persisted.proto_array_bytes)
+            .map_err(Error::InvalidProtoArrayBytes)?;
+
+        // Exit early if there are no "invalid" payloads, if requested.
+        if matches!(
+            reset_payload_statuses,
+            ResetPayloadStatuses::OnlyWithInvalidPayload
+        ) && !proto_array.contains_invalid_payloads()
+        {
+            return Ok(proto_array);
+        }
+
+        // Reset all blocks back to being "optimistic". This helps recover from an EL consensus
+        // fault where an invalid payload becomes valid.
```

Comment:
Can we add a debug log here so we know when this is being attempted, I think that might be useful

Reply:
Good call, I made that logger werk in https://github.com/sigp/lighthouse/pull/3498/commits/8551abf258ef37c068d4ce6f1e1d8fcb50e95264